packages/core/src/symbols/symbol-flow.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  Context,
+  effect,
+  getContext,
+  onCleanup,
+} from "@alloy-js/core/jsx-runtime";
+import { isRef, Ref, shallowReactive } from "@vue/reactivity";
+import { MemberScopeContext, ScopeContext } from "../index.browser.js";
+import { OutputSymbol, OutputSymbolFlags } from "./output-symbol.js";
+
+export interface TakeSymbolCallback {
+  (symbol: OutputSymbol): void;
+}
+
+export interface TakeSymbolsCallback {
+  (symbols: Set<OutputSymbol>): void;
+}
+
+export function takeSymbols() {
+  const context = getContext()!;
+  context.takesSymbols = true;
+  context.takenSymbols = shallowReactive(new Set<OutputSymbol>());
+  return context.takenSymbols;
+}
+
+export function emitSymbol(
+  symbol: OutputSymbol | Ref<OutputSymbol | undefined>,
+) {
+  let ownerContext: Context | undefined;
+  let context = getContext()!.owner;
+  while (context) {
+    if (context.takesSymbols) {
+      ownerContext = context;
+      break;
+    }
+
+    if (
+      context.context &&
+      (context.context[ScopeContext.id] ||
+        context.context[MemberScopeContext.id])
+    ) {
+      // don't cross scope boundaries.
+      break;
+    }
+
+    context = context.owner;
+  }
+
+  if (!ownerContext) {
+    return;
+  }
+
+  if (isRef(symbol)) {
+    effect<OutputSymbol | undefined>((prevSymbol) => {
+      onCleanup(() => {
+        if (symbol.value) {
+          ownerContext.takenSymbols!.delete(symbol.value);
+        }
+      });
+      if (symbol.value === undefined) {
+        ownerContext.takenSymbols!.delete(prevSymbol!);
+        return undefined;
+      } else {
+        ownerContext.takenSymbols!.add(symbol.value);
+        return symbol.value;
+      }
+    });
+  } else {
+    ownerContext.takenSymbols!.add(symbol);
+    onCleanup(() => {
+      context!.takenSymbols!.delete(symbol);
+    });
+  }
+}
+
+export function moveTakenMembersTo(baseSymbol: OutputSymbol) {
```

Comment:
The name of this function doesn't make it clear that it "activates" `takeSymbols`, I was initially confused about why the var declaration component was moving taken symbols but was not signing up for taking symbols first